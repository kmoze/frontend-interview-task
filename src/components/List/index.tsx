import React, { useEffect, useState } from 'react'
import { Allowance } from '@/types/allowances'
import { Inter } from 'next/font/google'
import Card from '@/components/Card'

const inter = Inter({ subsets: ['latin'] })

const List = () => {
  const [allowances, setAllowances] = useState<Allowance[]>([])

  useEffect(() => {
    const fetchData = async () => {
      const response = await fetch('/api/allowances')
      const { result } = await response.json()
      setAllowances(result)
    }

    fetchData()
  }, [])

  return (
    <div>
      <h2
        className={`${inter.className} text-2xl px-6 pt-6 font-semibold text-[#163B3B]`}
      >
        Allowances
      </h2>
      <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6 p-6">
        {allowances.map((allowance) => (
          <Card key={allowance.id} allowance={allowance} />
        ))}
      </div>
    </div>
  )
}

export default List
