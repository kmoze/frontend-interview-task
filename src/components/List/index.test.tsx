import { render, waitFor, screen } from '@testing-library/react'
import List from '@/components/List'
import '@testing-library/jest-dom'
import 'jest-fetch-mock'
import data from '@/data/data.json'

describe('List', () => {
  beforeEach(() => {
    fetchMock.mockResponse(JSON.stringify({ result: data }), { status: 200 })
  })

  it('renders allowances after fetching data', async () => {
    render(<List />)

    // Wait for at least one card to appear
    await waitFor(() => {
      expect(screen.getByText(data[0].name)).toBeInTheDocument()
    })

    // Check that the correct number of cards is rendered
    expect(screen.getAllByTestId('allowance-card').length).toBe(data.length)

    // Ensure some known allowances are present
    expect(screen.getByText('Learning & Development')).toBeInTheDocument()
    expect(screen.getByText('Wellbeing')).toBeInTheDocument()
    expect(screen.getByText('WFH')).toBeInTheDocument()
  })

  it('matches snapshot after data loads', async () => {
    const { container } = render(<List />)

    // Wait until the allowances are populated
    await waitFor(() => {
      expect(screen.getByText(data[0].name)).toBeInTheDocument()
    })

    expect(container).toMatchSnapshot()
  })
})
