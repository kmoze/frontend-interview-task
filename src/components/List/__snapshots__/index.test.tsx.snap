// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List matches snapshot after data loads 1`] = `
<div>
  <div>
    <h2
      class="className text-2xl px-6 pt-6 font-semibold text-[#163B3B]"
    >
      Allowances
    </h2>
    <div
      class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6 p-6"
    >
      <div
        class="group bg-white rounded-lg p-6 border border-[#DDDDDD] shadow-sm min-h-[185px] flex flex-col justify-between hover:shadow-md hover:scale-105 transition-transform duration-300"
        data-testid="allowance-card"
      >
        <div
          class="space-y-1.5"
        >
          <h3
            class="className text-base font-medium text-[#232323]"
          >
            Learning & Development
          </h3>
          <p
            class="className text-sm text-[#797979] font-normal"
          >
            Expense
          </p>
        </div>
        <div
          class="space-y-[-0.25rem]"
        >
          <div
            class="flex justify-between mb-2"
          >
            <span
              class="className text-[#232323] text-sm font-medium"
            >
              20
              % utilised
            </span>
            <span
              class="className text-[#9A9A9A] text-sm"
            >
              £
              100
               /
               
              Month
            </span>
          </div>
          <div
            class="w-full h-1 bg-[#DDDDDD] rounded-full"
          >
            <div
              class="h-full bg-[#64D98A] rounded-full"
              data-testid="progress-bar"
              style="width: 20%;"
            />
          </div>
        </div>
      </div>
      <div
        class="group bg-white rounded-lg p-6 border border-[#DDDDDD] shadow-sm min-h-[185px] flex flex-col justify-between hover:shadow-md hover:scale-105 transition-transform duration-300"
        data-testid="allowance-card"
      >
        <div
          class="space-y-1.5"
        >
          <h3
            class="className text-base font-medium text-[#232323]"
          >
            Wellbeing
          </h3>
          <p
            class="className text-sm text-[#797979] font-normal"
          >
            Spend Card
          </p>
        </div>
        <div
          class="space-y-[-0.25rem]"
        >
          <div
            class="flex justify-between mb-2"
          >
            <span
              class="className text-[#232323] text-sm font-medium"
            >
              93
              % utilised
            </span>
            <span
              class="className text-[#9A9A9A] text-sm"
            >
              £
              150
               /
               
              Month
            </span>
          </div>
          <div
            class="w-full h-1 bg-[#DDDDDD] rounded-full"
          >
            <div
              class="h-full bg-[#64D98A] rounded-full"
              data-testid="progress-bar"
              style="width: 93.33333333333333%;"
            />
          </div>
        </div>
      </div>
      <div
        class="group bg-white rounded-lg p-6 border border-[#DDDDDD] shadow-sm min-h-[185px] flex flex-col justify-between hover:shadow-md hover:scale-105 transition-transform duration-300"
        data-testid="allowance-card"
      >
        <div
          class="space-y-1.5"
        >
          <h3
            class="className text-base font-medium text-[#232323]"
          >
            WFH
          </h3>
          <p
            class="className text-sm text-[#797979] font-normal"
          >
            Spend Card
          </p>
        </div>
        <div
          class="space-y-[-0.25rem]"
        >
          <div
            class="flex justify-between mb-2"
          >
            <span
              class="className text-[#232323] text-sm font-medium"
            >
              80
              % utilised
            </span>
            <span
              class="className text-[#9A9A9A] text-sm"
            >
              £
              1000
               /
               
              Year
            </span>
          </div>
          <div
            class="w-full h-1 bg-[#DDDDDD] rounded-full"
          >
            <div
              class="h-full bg-[#64D98A] rounded-full"
              data-testid="progress-bar"
              style="width: 80%;"
            />
          </div>
        </div>
      </div>
      <div
        class="group bg-white rounded-lg p-6 border border-[#DDDDDD] shadow-sm min-h-[185px] flex flex-col justify-between hover:shadow-md hover:scale-105 transition-transform duration-300"
        data-testid="allowance-card"
      >
        <div
          class="space-y-1.5"
        >
          <h3
            class="className text-base font-medium text-[#232323]"
          >
            Lunch
          </h3>
          <p
            class="className text-sm text-[#797979] font-normal"
          >
            Spend Card
          </p>
        </div>
        <div
          class="-mx-6 -mb-6"
        >
          <div
            class="bg-[#FAFAFA] p-4 rounded-b-lg"
          >
            <button
              class="className text-[#569F6E] text-sm relative"
            >
              Activate card
              <span
                class="absolute left-0 bottom-0 w-0 h-[2px] bg-[#569F6E] transition-all duration-300 group-hover:w-full"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="group bg-white rounded-lg p-6 border border-[#DDDDDD] shadow-sm min-h-[185px] flex flex-col justify-between hover:shadow-md hover:scale-105 transition-transform duration-300"
        data-testid="allowance-card"
      >
        <div
          class="space-y-1.5"
        >
          <h3
            class="className text-base font-medium text-[#232323]"
          >
            Team Day
          </h3>
          <p
            class="className text-sm text-[#797979] font-normal"
          >
            Spend Card
          </p>
        </div>
        <div
          class="-mx-6 -mb-6"
        >
          <div
            class="bg-[#FAFAFA] p-4 rounded-b-lg"
          >
            <button
              class="className text-[#569F6E] text-sm relative"
            >
              Activate card
              <span
                class="absolute left-0 bottom-0 w-0 h-[2px] bg-[#569F6E] transition-all duration-300 group-hover:w-full"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="group bg-white rounded-lg p-6 border border-[#DDDDDD] shadow-sm min-h-[185px] flex flex-col justify-between hover:shadow-md hover:scale-105 transition-transform duration-300"
        data-testid="allowance-card"
      >
        <div
          class="space-y-1.5"
        >
          <h3
            class="className text-base font-medium text-[#232323]"
          >
            Birthday
          </h3>
          <p
            class="className text-sm text-[#797979] font-normal"
          >
            Spend Card
          </p>
        </div>
        <div
          class="space-y-[-0.25rem]"
        >
          <div
            class="flex justify-between mb-2"
          >
            <span
              class="className text-[#232323] text-sm font-medium"
            >
              100
              % utilised
            </span>
            <span
              class="className text-[#9A9A9A] text-sm"
            >
              £
              100
               /
               
              Year
            </span>
          </div>
          <div
            class="w-full h-1 bg-[#DDDDDD] rounded-full"
          >
            <div
              class="h-full bg-[#64D98A] rounded-full"
              data-testid="progress-bar"
              style="width: 100%;"
            />
          </div>
        </div>
      </div>
      <div
        class="group bg-white rounded-lg p-6 border border-[#DDDDDD] shadow-sm min-h-[185px] flex flex-col justify-between hover:shadow-md hover:scale-105 transition-transform duration-300"
        data-testid="allowance-card"
      >
        <div
          class="space-y-1.5"
        >
          <h3
            class="className text-base font-medium text-[#232323]"
          >
            Software
          </h3>
          <p
            class="className text-sm text-[#797979] font-normal"
          >
            Spend Card
          </p>
        </div>
        <div
          class="space-y-[-0.25rem]"
        >
          <div
            class="flex justify-between mb-2"
          >
            <span
              class="className text-[#232323] text-sm font-medium"
            >
              30
              % utilised
            </span>
            <span
              class="className text-[#9A9A9A] text-sm"
            >
              £
              30
               /
               
              Month
            </span>
          </div>
          <div
            class="w-full h-1 bg-[#DDDDDD] rounded-full"
          >
            <div
              class="h-full bg-[#64D98A] rounded-full"
              data-testid="progress-bar"
              style="width: 29.96666666666667%;"
            />
          </div>
        </div>
      </div>
      <div
        class="group bg-white rounded-lg p-6 border border-[#DDDDDD] shadow-sm min-h-[185px] flex flex-col justify-between hover:shadow-md hover:scale-105 transition-transform duration-300"
        data-testid="allowance-card"
      >
        <div
          class="space-y-1.5"
        >
          <h3
            class="className text-base font-medium text-[#232323]"
          >
            Team Budget
          </h3>
          <p
            class="className text-sm text-[#797979] font-normal"
          >
            Spend Card
          </p>
        </div>
        <div
          class="space-y-[-0.25rem]"
        >
          <div
            class="flex justify-between mb-2"
          >
            <span
              class="className text-[#232323] text-sm font-medium"
            >
              31
              % utilised
            </span>
            <span
              class="className text-[#9A9A9A] text-sm"
            >
              £
              400
               /
               
              Year
            </span>
          </div>
          <div
            class="w-full h-1 bg-[#DDDDDD] rounded-full"
          >
            <div
              class="h-full bg-[#64D98A] rounded-full"
              data-testid="progress-bar"
              style="width: 31.25%;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
